moshi/src/main/java/com/squareup/moshi/Moshi.java

```diff
@@ -283,38 +245,128 @@ Builder addAll(List<JsonAdapter.Factory> factories) {
   }
 
   /**
-   * Sometimes a type adapter factory depends on its own product; either directly or indirectly.
-   * To make this work, we offer this type adapter stub while the final adapter is being computed.
-   * When it is ready, we wire this to delegate to that finished adapter.
+   * A possibly-reentrant call to lookup a JSON adapter.
+   *
+   * <p>We keep track of the current stack of lookups: we may start by looking up the JSON adapter
+   * for Employee, re-enter looking for the JSON adapter of HomeAddress, and re-enter again looking
+   * up the JSON adapter of PostalCode. If any of these lookups fail we can provide a stack trace
+   * with all of the lookups.
+   *
+   * <p>Sometimes a JSON adapter factory depends on its own product; either directly or indirectly.
+   * To make this work, we offer a JSON adapter stub while the final adapter is being computed.
+   * When it is ready, we wire the stub to that finished adapter. This is necessary in
+   * self-referential object models, such as an {@code Employee} class that has a {@code
+   * List<Employee>} field for an organization's management hierarchy.
    *
-   * <p>Typically this is necessary in self-referential object models, such as an {@code Employee}
-   * class that has a {@code List<Employee>} field for an organization's management hierarchy.
+   * <p>This class defers putting any JSON adapters in the cache until the topmost JSON adapter has
+   * successfully been computed. That way we don't pollute the cache with incomplete stubs, or
+   * adapters that may transitively depend on incomplete stubs.
    */
-  private static final class DeferredAdapter<T> extends JsonAdapter<T> {
+  final class Call {
```

Comment:
If only to avoid confusion for having a third Call in our lives, is there anything else we can call this? don't have any good suggestions though...

Reply:
First name was ReentrantCall. Maybe LookupRoot ?

Reply:
LookupChain? LookupTree?

Reply:
What about reverting `Lookup` to `DeferredAdapter` and naming this as the (stateful) `Lookup`?

Otherwise root and chain both seem fine.

Reply:
Renamed to LookupChain!